lib/chewy/type/syncer.rb

```diff
@@ -19,10 +19,79 @@ class Type
     # @see Chewy::Type::Actions::ClassMethods#sync
     class Syncer
       DEFAULT_SYNC_BATCH_SIZE = 20_000
+      ISO_DATETIME = /\A(\d{4})-(\d\d)-(\d\d) (\d\d):(\d\d):(\d\d)(\.\d+)?\z/
+      OUTDATED_IDS_WORKER = lambda do |outdated_sync_field_type, source_data_hash, index_data|
+        time = Time.now
+        r = index_data.each_with_object([]) do |(id, index_sync_value), result|
+          next unless source_data_hash[id]
+
+          outdated = if outdated_sync_field_type == 'date'
+            !Chewy::Type::Syncer.dates_equal(typecast_date(source_data_hash[id]), DateTime.iso8601(index_sync_value))
+          else
+            source_data_hash[id] != index_sync_value
+          end
+
+          result.push(id) if outdated
+        end
+        p "Batch processing time: #{Time.now - time}"
+        r
+      end
+      SOURCE_OR_INDEX_DATA_WORKER = lambda do |syncer, type|
+        result = case type
+        when :source
+          require 'ruby-prof'
+          r = nil
+          p(Benchmark.measure('source') do
+            RubyProf.start
+            r = syncer.send(:fetch_source_data)
+            result = RubyProf.stop
+            printer = RubyProf::GraphPrinter.new(result)
+            # printer.print(STDOUT)
+          end)
+          r
+        when :index
+          r = nil
+          p(Benchmark.measure('index') do
+            r = syncer.send(:fetch_index_data)
+          end)
+          r
+        end
+        {type => result}
+      end
+
+      def self.typecast_date(string)
+        if string.is_a?(String) && string =~ ISO_DATETIME
+          microsec = ($7.to_r * 1_000_000).to_i
+          date = "#{$1}-#{$2}-#{$3}T#{$4}:#{$5}:#{$6}.#{'%06d' % microsec}+00:00"
+          DateTime.iso8601(date)
+        else
+          string
+        end
+      end
+
+      # Compares times with ms precision.
+      def self.dates_equal(one, two)
+        [one.to_i, one.usec / 1000] == [two.to_i, two.usec / 1000]
+      end
+
+      # In ActiveSupport ~> 4.0 json dumpled times without any
+      # milliseconds, so ES stored time with the seconds precision.
+      if ActiveSupport::VERSION::STRING < '4.1.0'
+        def self.dates_equal(one, two)
+          one.to_i == two.to_i
+        end
+      end
 
       # @param type [Chewy::Type] chewy type
-      def initialize(type)
+      def initialize(type, parallel: 4)
```

Comment:
it will be parallel: nil and will be passed from the rake helper, I just didn't implement the interface yet